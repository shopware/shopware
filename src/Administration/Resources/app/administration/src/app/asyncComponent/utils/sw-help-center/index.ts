import template from './sw-help-center.html.twig';
import './sw-help-center.scss';

/**
 * @description Displays an icon and a link to the help sidebar
 *
 * @package buyers-experience
 *
 * @private
 */
export default Shopware.Component.wrapComponentConfig({
    template,

    computed: {
        showHelpSidebar(): boolean {
            return Shopware.State.get('adminHelpCenter').showHelpSidebar;
        },

        showShortcutModal(): boolean {
            return Shopware.State.get('adminHelpCenter').showShortcutModal;
        },
    },

    watch: {
        showShortcutModal(value) {
            const shortcutModal = this.$refs.shortcutModal as { onOpenShortcutOverviewModal: () => void };

            if (!shortcutModal) {
                return;
            }

            if (value === false) {
                this.setFocusToSidebar();

                return;
            }

            shortcutModal.onOpenShortcutOverviewModal();
        },
    },

    methods: {
        openHelpSidebar(): void {
            Shopware.State.commit('adminHelpCenter/setShowHelpSidebar', true);
        },

        openShortcutModal(): void {
            Shopware.State.commit('adminHelpCenter/setShowShortcutModal', true);
        },

        closeShortcutModal(): void {
            Shopware.State.commit('adminHelpCenter/setShowShortcutModal', false);
        },

        setFocusToSidebar(): void {
            const helpSidebar = this.$refs.helpSidebar as { setFocusToSidebar: () => void };

            if (!helpSidebar) {
                return;
            }

            helpSidebar.setFocusToSidebar();
        },
    },
});
