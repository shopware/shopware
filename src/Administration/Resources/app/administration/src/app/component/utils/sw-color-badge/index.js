import template from './sw-color-badge.html.twig';
import './sw-color-badge.scss';

const { Component } = Shopware;

/**
 * @public
 * @description
 * Renders a colored badge for example as indicator if an item is available.
 * @status ready
 * @example-type static
 * @component-example
 * <div>
 *     <!-- red square -->
 *     <sw-color-badge color="red"></sw-color-badge>
 *     <!-- green square -->
 *     <sw-color-badge color="green"></sw-color-badge>
 *     <!-- red circle -->
 *     <sw-color-badge color="red" rounded></sw-color-badge>
 * </div>
 */
Component.register('sw-color-badge', {
    template,

    props: {
        variant: {
            type: String,
            required: false,
            default: 'default'
        },
        color: {
            type: String,
            required: false,
            default: ''
        },
        rounded: {
            type: Boolean,
            required: false,
            default: false
        }
    },

    computed: {
        colorStyle() {
            if (!this.color.length) {
                return '';
            }
            return `background:${this.color}`;
        },
        variantClass() {
            return {
                [`is--${this.variant}`]: true,
                'is--rounded': this.rounded
            };
        }
    }
});
