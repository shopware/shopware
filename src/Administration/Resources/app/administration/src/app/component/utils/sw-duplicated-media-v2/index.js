import template from './sw-duplicated-media-v2.html.twig';
import './sw-duplicated-media-v2.scss';

const { Component, Context, Filter } = Shopware;
const { Criteria } = Shopware.Data;

/**
 * @private
 */

const LOCAL_STORAGE_KEY_OPTION = 'sw-duplicate-media-resolve-option';
const LOCAL_STORAGE_SAVE_SELECTION = 'sw-duplicate-media-resolve-save-selection';

Component.register('sw-duplicated-media-v2', {
    template,

    inject: ['repositoryFactory', 'mediaService'],

    data() {
        return {
            isLoading: false,
            shouldSaveSelection: false,
            selectedOption: 'Replace',
            suggestedName: '',
            existingMedia: null,
            targetEntity: null,
            failedUploadTasks: [],
            postponedFailedUploads: [],
        };
    },

    computed: {
        mediaRepository() {
            return this.repositoryFactory.create('media');
        },

        additionalErrorCount() {
            return this.failedUploadTasks.length - 1;
        },

        hasAdditionalErrors() {
            return this.additionalErrorCount > 0;
        },

        currentTask() {
            return this.failedUploadTasks[0];
        },

        buttonLabel() {
            return this.$tc(`global.sw-duplicated-media-v2.button${this.selectedOption}`);
        },

        dateFilter() {
            return Filter.getByName('date');
        },

        fileSizeFilter() {
            return Filter.getByName('fileSize');
        },

        currentTaskDetails() {
            if (!this.currentTask) {
                return '';
            }
            const metadata = [
                this.dateFilter(new Date(), { month: 'long' }),
            ];

            if (this.currentTask.src instanceof File) {
                metadata.push(this.fileSizeFilter(this.currentTask.src.size));
            }

            return metadata.join(', ');
        },

        showModal() {
            return this.failedUploadTasks.length > 0 && !this.isWorkingOnMultipleTasks;
        },

        isWorkingOnMultipleTasks() {
            return this.isLoading && this.shouldSaveSelection;
        },

        options() {
            return [
                {
                    value: 'Replace',
                    name: this.$tc('global.sw-duplicated-media-v2.labelOptionReplace'),
                },
                {
                    value: 'Rename',
                    name: this.$tc('global.sw-duplicated-media-v2.labelOptionRename'),
                },
                {
                    value: 'Skip',
                    name: this.$tc('global.sw-duplicated-media-v2.labelOptionSkip'),
                },
            ];
        },
    },

    watch: {
        currentTask() {
            this.updatePreviewData();
        },

        showModal(newVal) {
            if (newVal) {
                this.loadDefaultOption();
                return;
            }

            this.saveDefaultOption();
        },

        isLoading(newVal) {
            if (newVal) {
                return;
            }

            this.failedUploadTasks.push(...this.postponedFailedUploads.splice(0, this.postponedFailedUploads.length));
        },
    },

    created() {
        this.createdComponent();
    },

    beforeDestroy() {
        this.beforeDestroyComponent();
    },

    methods: {
        createdComponent() {
            this.loadDefaultOption();
            this.updatePreviewData();

            this.mediaService.addDefaultListener(this.handleMediaServiceUploadEvent);
        },

        beforeDestroyComponent() {
            this.mediaService.removeDefaultListener(this.handleMediaServiceUploadEvent);
        },

        loadDefaultOption() {
            this.shouldSaveSelection = localStorage.getItem(LOCAL_STORAGE_SAVE_SELECTION) || false;

            if (this.shouldSaveSelection) {
                this.defaultOption = localStorage.getItem(LOCAL_STORAGE_KEY_OPTION) || 'Replace';
            }

            this.selectedOption = this.defaultOption || 'Replace';
        },

        saveDefaultOption() {
            localStorage.setItem(LOCAL_STORAGE_SAVE_SELECTION, this.shouldSaveSelection);

            if (this.shouldSaveSelection) {
                localStorage.setItem(LOCAL_STORAGE_KEY_OPTION, this.defaultOption);
            }
        },

        handleMediaServiceUploadEvent({ action, payload }) {
            if (action !== 'media-upload-fail') {
                return;
            }

            if (!this.isDuplicatedNameError(payload.error)) {
                return;
            }

            if (this.isLoading) {
                this.postponedFailedUploads.push(payload);
                return;
            }

            this.failedUploadTasks.push(payload);
        },

        isDuplicatedNameError(error) {
            return error.response.data.errors.some((err) => {
                return err.code === 'CONTENT__MEDIA_DUPLICATED_FILE_NAME';
            });
        },

        async updatePreviewData() {
            if (!this.currentTask) {
                this.existingMedia = null;
                this.suggestedName = '';
                return;
            }

            const criteria = new Criteria(1, 1)
                .addFilter(
                    Criteria.multi(
                        'AND',
                        [
                            Criteria.equals('fileName', this.currentTask.fileName),
                            Criteria.equals('fileExtension', this.currentTask.extension),
                        ],
                    ),
                );

            const searchResult = await this.mediaRepository.search(criteria, Context.api);
            if (searchResult?.[0]) {
                this.existingMedia = searchResult[0];
            }

            const provided = await this.mediaService.provideName(this.currentTask.fileName, this.currentTask.extension);
            this.suggestedName = provided.fileName;
        },

        solveDuplicate() {
            if (!this.currentTask) {
                this.isLoading = false;
                return;
            }

            this.isLoading = true;

            switch (this.selectedOption) {
                case 'Rename':
                    this.renameFile(this.currentTask);
                    break;
                case 'Replace':
                    this.replaceFile(this.currentTask);
                    break;
                case 'Skip':
                default:
                    this.skipFile(this.currentTask);
                    break;
            }

            this.failedUploadTasks.splice(0, 1);

            if (!this.currentTask) {
                this.isLoading = false;
            } else {
                this.solveDuplicate();
            }
        },

        async renameFile(uploadTask) {
            const newTask = Object.assign({}, uploadTask);

            const { fileName } = await this.mediaService.provideName(uploadTask.fileName, uploadTask.extension);
            newTask.fileName = fileName;

            this.mediaService.addUpload(newTask.uploadTag, newTask);
            await this.mediaService.runUploads(newTask.uploadTag);
        },

        skipAll() {
            this.isLoading = true;

            this.skipFile(this.currentTask);
            this.failedUploadTasks.splice(0, 1);

            if (!this.currentTask) {
                this.isLoading = false;
            } else {
                this.skipAll();
            }
        },

        skipCurrentFile() {
            this.isLoading = true;
            this.skipFile(this.currentTask);

            this.failedUploadTasks.splice(0, 1);
            this.isLoading = false;
        },

        async skipFile(uploadTask) {
            const oldTarget = await this.mediaRepository.get(uploadTask.targetId, Context.api);
            if (!oldTarget.hasFile) {
                await this.mediaRepository.delete(oldTarget.id, Context.api);
            }

            this.mediaService.cancelUpload(uploadTask.uploadTag, uploadTask);
        },

        async replaceFile(uploadTask) {
            const criteria = new Criteria(1, 1)
                .addFilter(Criteria.multi('AND',
                    [
                        Criteria.equals('fileName', uploadTask.fileName),
                        Criteria.equals('fileExtension', uploadTask.extension),
                    ]));

            const searchResult = await this.mediaRepository.search(criteria, Context.api);
            const newTarget = searchResult[0];
            const oldTargetId = uploadTask.targetId;
            uploadTask.targetId = newTarget.id;

            this.mediaService.addUpload(uploadTask.uploadTag, uploadTask);

            await this.mediaService.runUploads(uploadTask.uploadTag);
            const oldTarget = await this.mediaRepository.get(oldTargetId, Context.api);

            if (!oldTarget.hasFile) {
                await this.mediaRepository.delete(oldTargetId, Context.api);
            }

            await this.mediaRepository.get(uploadTask.targetId, Context.api);
        },
    },
});
