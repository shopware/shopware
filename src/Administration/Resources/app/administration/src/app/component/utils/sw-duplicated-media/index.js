import CriteriaFactory from 'src/core/factory/criteria.factory';
import template from './sw-duplicated-media.html.twig';
import './sw-duplicated-media.scss';

const { Component, StateDeprecated, Filter } = Shopware;

/**
 * @private
 */

const LOCAL_STORAGE_KEY_OPTION = 'sw-duplicate-media-resolve-option';
const LOCAL_STORAGE_SAVE_SELECTION = 'sw-duplicate-media-resolve-save-selection';

Component.register('sw-duplicated-media', {
    template,

    inject: [
        'mediaService'
    ],

    data() {
        return {
            isLoading: false,
            shouldSaveSelection: false,
            selectedOption: 'Replace',
            suggestedName: '',
            existingMedia: null,
            targetEntity: null,
            failedUploadTasks: [],
            postponedFailedUploads: []
        };
    },

    computed: {
        mediaStore() {
            return StateDeprecated.getStore('media');
        },

        uploadStore() {
            return StateDeprecated.getStore('upload');
        },

        additionalErrorCount() {
            return this.failedUploadTasks.length - 1;
        },

        hasAdditionalErrors() {
            return this.additionalErrorCount > 0;
        },

        currentTask() {
            return this.failedUploadTasks[0];
        },

        buttonLabel() {
            return this.$tc(`global.sw-duplicated-media.button${this.selectedOption}`);
        },

        dateFilter() {
            return Filter.getByName('date');
        },

        fileSizeFilter() {
            return Filter.getByName('fileSize');
        },

        currentTaskDetails() {
            if (!this.currentTask) {
                return '';
            }
            const metadata = [
                this.dateFilter(new Date(), { month: 'long' })
            ];

            if (this.currentTask.src instanceof File) {
                metadata.push(this.fileSizeFilter(this.currentTask.src.size));
            }

            return metadata.join(', ');
        },

        showModal() {
            return this.failedUploadTasks.length > 0 && !this.isWorkingOnMultipleTasks;
        },

        isWorkingOnMultipleTasks() {
            return this.isLoading && this.shouldSaveSelection;
        },

        options() {
            return [
                {
                    value: 'Replace',
                    name: this.$tc('global.sw-duplicated-media.labelOptionReplace')
                },
                {
                    value: 'Rename',
                    name: this.$tc('global.sw-duplicated-media.labelOptionRename')
                },
                {
                    value: 'Skip',
                    name: this.$tc('global.sw-duplicated-media.labelOptionSkip')
                }
            ];
        }
    },

    watch: {
        currentTask() {
            this.updatePreviewData();
        },

        showModal(newVal) {
            if (newVal) {
                this.loadDefaultOption();
                return;
            }

            this.saveDefaultOption();
        },

        isLoading(newVal) {
            if (newVal) {
                return;
            }

            this.failedUploadTasks.push(...this.postponedFailedUploads.splice(0, this.postponedFailedUploads.length));
        }
    },

    created() {
        this.createdComponent();
    },

    beforeDestroy() {
        this.beforeDestroyComponent();
    },

    methods: {
        createdComponent() {
            this.loadDefaultOption();
            this.updatePreviewData();

            this.uploadStore.addDefaultListener(this.handleUploadStoreEvent);
        },

        beforeDestroyComponent() {
            this.uploadStore.removeDefaultListener(this.handleUploadStoreEvent);
        },

        loadDefaultOption() {
            this.shouldSaveSelection = localStorage.getItem(LOCAL_STORAGE_SAVE_SELECTION) || false;
            if (this.shouldSaveSelection) {
                this.defaultOption = localStorage.getItem(LOCAL_STORAGE_KEY_OPTION) || 'Replace';
            }
            this.selectedOption = this.defaultOption || 'Replace';
        },

        saveDefaultOption() {
            localStorage.setItem(LOCAL_STORAGE_SAVE_SELECTION, this.shouldSaveSelection);
            if (this.shouldSaveSelection) {
                localStorage.setItem(LOCAL_STORAGE_KEY_OPTION, this.defaultOption);
            }
        },

        handleUploadStoreEvent({ action, payload }) {
            if (action !== 'media-upload-fail') {
                return;
            }

            if (!this.isDuplicatedNameError(payload.error)) {
                return;
            }

            if (this.isLoading) {
                this.postponedFailedUploads.push(payload);
                return;
            }

            this.failedUploadTasks.push(payload);
        },

        isDuplicatedNameError(error) {
            return error.response.data.errors.some((err) => {
                return err.code === 'CONTENT__MEDIA_DUPLICATED_FILE_NAME';
            });
        },

        updatePreviewData() {
            if (!this.currentTask) {
                this.existingMedia = null;
                this.suggestedName = '';
                return;
            }

            this.mediaStore.getList({
                page: 1,
                limit: 1,
                criteria: CriteriaFactory.multi('AND',
                    CriteriaFactory.equals('fileName', this.currentTask.fileName),
                    CriteriaFactory.equals('fileExtension', this.currentTask.extension))
            }).then((response) => {
                this.existingMedia = response.items[0];
            });

            this.mediaService.provideName(this.currentTask.fileName, this.currentTask.extension).then((response) => {
                this.suggestedName = response.fileName;
            });
        },

        solveDuplicate() {
            if (!this.currentTask) {
                this.isLoading = false;
                return Promise.resolve();
            }

            this.isLoading = true;
            let solvingPromise = null;
            switch (this.selectedOption) {
                case 'Rename':
                    solvingPromise = this.renameFile(this.currentTask);
                    break;
                case 'Replace':
                    solvingPromise = this.replaceFile(this.currentTask);
                    break;
                case 'Skip':
                default:
                    solvingPromise = this.skipFile(this.currentTask);
                    break;
            }

            return solvingPromise.then(() => {
                this.failedUploadTasks.splice(0, 1);
                if (this.shouldSaveSelection) {
                    return this.solveDuplicate();
                }
                this.isLoading = false;
                return Promise.resolve();
            }).catch((cause) => {
                this.isLoading = false;
                return Promise.reject(cause);
            });
        },

        renameFile(uploadTask) {
            const newTask = Object.assign({}, uploadTask);
            return this.mediaService.provideName(uploadTask.fileName, uploadTask.extension).then(({ fileName }) => {
                newTask.fileName = fileName;
                this.uploadStore.addUpload(newTask.uploadTag, newTask);
                return this.uploadStore.runUploads(newTask.uploadTag);
            });
        },

        skipAll() {
            this.isLoading = true;

            return this.skipFile(this.currentTask).then(() => {
                this.failedUploadTasks.splice(0, 1);
                if (this.currentTask) {
                    return this.skipAll();
                }

                this.isLoading = false;
                return Promise.resolve();
            }).catch(() => {
                this.isLoading = false;
                return Promise.reject();
            });
        },

        skipCurrentFile() {
            this.isLoading = true;
            this.skipFile(this.currentTask).then(() => {
                this.failedUploadTasks.splice(0, 1);
                this.isLoading = false;
            }).catch(() => {
                this.isLoading = false;
            });
        },

        skipFile(uploadTask) {
            return this.mediaStore.getByIdAsync(uploadTask.targetId).then((oldTarget) => {
                if (!oldTarget.hasFile) {
                    oldTarget.delete(true);
                }
            });
        },

        replaceFile(uploadTask) {
            return this.mediaStore.getList({
                page: 1,
                limit: 1,
                criteria: CriteriaFactory.multi('AND',
                    CriteriaFactory.equals('fileName', uploadTask.fileName),
                    CriteriaFactory.equals('fileExtension', uploadTask.extension))
            }).then((response) => {
                const newTarget = response.items[0];
                const oldTargetId = uploadTask.targetId;
                uploadTask.targetId = newTarget.id;

                this.uploadStore.addUpload(uploadTask.uploadTag, uploadTask);
                return this.uploadStore.runUploads(uploadTask.uploadTag).then(() => {
                    return this.mediaStore.getByIdAsync(oldTargetId).then((oldTarget) => {
                        if (!oldTarget.hasFile) {
                            oldTarget.delete(true);
                        }
                        return this.mediaStore.getByIdAsync(uploadTask.targetId);
                    });
                });
            });
        }
    }
});
