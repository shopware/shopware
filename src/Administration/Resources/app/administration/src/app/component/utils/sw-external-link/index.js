import template from './sw-external-link.html.twig';
import './sw-external-link.scss';

const { Component } = Shopware;

/**
 * @public
 * @description Link to another website outside the admin, that opens in a new browser tab
 * @status ready
 * @example-type dynamic
 * @component-example
 * <sw-external-link
 *   href="https://google.com">
 *   Ask google
 * </sw-external-link>
 */
Component.register('sw-external-link', {
    template,

    inheritAttrs: false,

    props: {
        small: {
            type: Boolean,
            required: false,
            default: false,
        },

        icon: {
            type: String,
            required: false,
            default: 'small-arrow-small-external',
        },
    },

    computed: {
        classes() {
            return {
                'sw-external-link--small': this.small,
            };
        },

        iconSize() {
            if (this.small) {
                return '8px';
            }

            return '10px';
        },
    },

    methods: {
        onClick(event) {
            this.$emit('click', event);
        },
    },
});
