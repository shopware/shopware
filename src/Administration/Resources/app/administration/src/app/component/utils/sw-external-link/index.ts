import template from './sw-external-link.html.twig';

const { Component } = Shopware;

/**
 * @package admin
 *
 * @private
 * @status ready
 * @description Wrapper component for sw-external-link and mt-external-link. Autoswitches between the two components.
 */
Component.register('sw-external-link', {
    template,

    compatConfig: Shopware.compatConfig,

    computed: {
        useMeteorComponent() {
            // Use new meteor component in major
            if (Shopware.Feature.isActive('v6.7.0.0')) {
                return true;
            }

            // Throw warning when deprecated component is used
            Shopware.Utils.debug.warn(
                'sw-external-link',
                // eslint-disable-next-line max-len
                'The old usage of "sw-external-link" is deprecated and will be removed in v6.7.0.0. Please use "mt-external-link" instead.',
            );

            return false;
        },

        listeners() {
            if (this.isCompatEnabled('INSTANCE_LISTENERS')) {
                return this.$listeners;
            }

            return {};
        },
    },

    methods: {
        getSlots() {
            // eslint-disable-next-line @typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-member-access
            if (this.isCompatEnabled('INSTANCE_SCOPED_SLOTS')) {
                return {
                    ...this.$slots,
                    ...this.$scopedSlots,
                };
            }

            return this.$slots;
        },
    },
});
