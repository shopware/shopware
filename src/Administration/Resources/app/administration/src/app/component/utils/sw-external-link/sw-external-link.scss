@import "~scss/variables";

.sw-external-link {
    color: $color-shopware-brand-500;
    font-size: $font-size-small;
    text-decoration: underline;
    cursor: pointer;

    &:hover,
    &:focus {
        color: $color-shopware-brand-600;
    }

    &__icon {
        margin-left: 4px;
    }

    &--small {
        font-size: $font-size-extra-small;

        .sw-external-link__icon {
            margin-left: 2px;
        }
    }
}
