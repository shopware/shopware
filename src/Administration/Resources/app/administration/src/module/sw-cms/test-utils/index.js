import { mount } from '@vue/test-utils';

function runGenericCmsTest(component) {
    beforeAll(async () => {
        await setupCmsEnvironment();
    });

    it('should be a Vue.js component', async () => {
        const wrapper = mount(component);
        expect(wrapper.vm).toBeTruthy();
    });
}

function runCmsBlockRegistryTest(config) {
    beforeAll(async () => {
        await setupCmsEnvironment();
        await import(config.import);
    });

    it('registers the component in the CMS block registry', async () => {
        const componentRegistry = Shopware.Component.getComponentRegistry();
        const cmsService = Shopware.Service('cmsService');

        expect(componentRegistry.has(config.component)).toBeTruthy();
        expect(componentRegistry.has(config.preview)).toBeTruthy();

        const blockConfig = cmsService.getCmsBlockConfigByName(config.name);

        expect(blockConfig.component).toBe(config.component);
        expect(blockConfig.previewComponent).toBe(config.preview);
    });
}

function runCmsElementRegistryTest(config) {
    beforeAll(async () => {
        await setupCmsEnvironment();
        await import(config.import);
    });

    it('registers the component in the CMS block registry', async () => {
        const componentRegistry = Shopware.Component.getComponentRegistry();
        const cmsService = Shopware.Service('cmsService');

        expect(componentRegistry.has(config.component)).toBeTruthy();

        if (config.preview) {
            expect(componentRegistry.has(config.preview)).toBeTruthy();
        }

        const elementConfig = cmsService.getCmsElementConfigByName(config.name);

        expect(elementConfig.component).toBe(config.component);
        expect(elementConfig.configComponent).toBe(config.config);

        if (config.preview) {
            expect(elementConfig.previewComponent).toBe(config.preview);
        }
    });
}

async function setupCmsEnvironment() {
    await import('src/module/sw-cms/store/cms-page.store');
    await import('src/module/sw-cms/service/cms.service');
    await import('src/module/sw-cms/service/cms-element-favorites.service');
    await import('src/module/sw-cms/mixin/sw-cms-element.mixin');
    await import('src/module/sw-cms/mixin/sw-cms-state.mixin');

    Shopware.State.get('session').currentUser = {
        id: 'admin',
    };
}

/**
 * @private
 * @package buyers-experience
 */
export { runGenericCmsTest, runCmsBlockRegistryTest, runCmsElementRegistryTest, setupCmsEnvironment };
