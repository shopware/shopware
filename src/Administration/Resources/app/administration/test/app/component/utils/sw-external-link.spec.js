import { shallowMount } from '@vue/test-utils';
import 'src/app/component/utils/sw-external-link';

// initial component setup
const setup = (propOverride) => {
    const propsData = {
        ...propOverride
    };

    return shallowMount(Shopware.Component.build('sw-external-link'), {
        stubs: ['sw-icon'],
        slots: {
            default: 'test external link'
        },
        propsData
    });
};

describe('components/utils/sw-external-link', () => {
    it('should be a Vue.js component', () => {
        const wrapper = setup({ href: 'https://google.com' });
        expect(wrapper.vm).toBeTruthy();
    });

    it('should render correctly', () => {
        const wrapper = setup({ href: 'https://google.com' });
        expect(wrapper.element).toMatchSnapshot();
    });

    it('should display a custom icon', () => {
        const wrapper = setup({
            href: 'https://google.com',
            icon: 'default-test-icon'
        });

        expect(wrapper.find('sw-icon-stub').attributes().name).toBe('default-test-icon');
        expect(wrapper.find('sw-icon-stub').attributes().size).toBe('10px');
    });

    it('should emit click event if no href is provided', async () => {
        const wrapper = setup();

        wrapper.trigger('click');
        await wrapper.vm.$nextTick();

        expect(wrapper.emitted().click).toBeTruthy();
    });

    it('should render small', async () => {
        const wrapper = setup({
            href: 'https://google.com',
            small: true
        });

        expect(wrapper.find('sw-icon-stub').attributes().size).toBe('8px');

        expect(wrapper.classes()).toContain('sw-external-link--small');
    });
});
