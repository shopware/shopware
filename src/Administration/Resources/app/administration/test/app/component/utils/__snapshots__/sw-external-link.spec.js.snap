// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/utils/sw-external-link should render correctly 1`] = `
<a
  class="sw-external-link"
  href="https://google.com"
  rel="noopener"
  target="_blank"
>
  test external link 
  <sw-icon-stub
    class="sw-external-link__icon"
    name="regular-external-link-s"
    size="10px"
  />
</a>
`;
